actionpack/lib/action_dispatch/middleware/server_timing.rb

```diff
@@ -6,30 +6,72 @@ module ActionDispatch
   class ServerTiming
     SERVER_TIMING_HEADER = "Server-Timing"
 
+    class Subscriber # :nodoc:
+      include Singleton
+      KEY = :action_dispatch_server_timing_events
+
+      def initialize
+        @mutex = Mutex.new
+      end
+
+      def call(event)
+        if events = ActiveSupport::IsolatedExecutionState[KEY]
+          events << event
+        end
+      end
+
+      def collect_events
+        events = []
+        ActiveSupport::IsolatedExecutionState[KEY] = events
+        yield
+        events
+      ensure
+        ActiveSupport::IsolatedExecutionState.delete(KEY)
+      end
+
+      def ensure_subscribed
+        @mutex.synchronize do
+          # Subscribe to all events, except those beginning with "!"
+          # Ideally we would be more selective of what is being measured
+          @subscriber ||= ActiveSupport::Notifications.subscribe(/\A[^!]/, self)
+        end
+      end
+
+      def unsubscribe
+        @mutex.synchronize do
+          ActiveSupport::Notifications.unsubscribe @subscriber
+          @subscriber = nil
+        end
+      end
+    end
+
+    def self.unsubscribe # :nodoc:
+      Subscriber.instance.unsubscribe
+    end
+
     def initialize(app)
       @app = app
     end
 
     def call(env)
-      events = []
-      subscriber = ActiveSupport::Notifications.subscribe(/.*/) do |*args|
-        events << ActiveSupport::Notifications::Event.new(*args)
-      end
+      subscriber = Subscriber.instance
+      subscriber.ensure_subscribed
```

Comment:
We want to be able to `unsubscribe` in test teardown (after a test that has initialized / subscribed), so the subscription doesn't hang around... but the singleton still needs to then re-subscribe in a future test later when needed.

Reply:
I think you could do both by having a `Subscriber.subscribe` method. But since it's intended as a dev only middleware it likely don't matter.

Reply:
Hmm. I do somewhat want to subscribe earlier, if for no other reason than that it will likely mitigate issues with the subscriber being dynamic (I would like to backport this to 7.0.x)

The hesitation I have is that `unsubscribe` then breaks any ServerTiming instance that is active (though I guess we aren't expecting any others to be active in the test suite, which is the only place we'll be calling unsubscribe)